### ISSUE TITLE
question: how would i annotate a relation that i only want to lookup if included

### ISSUE THREAD

**Issue by jmchuster** [crnk-project/crnk-framework#197]
For, example, let's say events have associated users.
So, we have
```
@JsonApiResource(type='events')
public class EventResource {
  @JsonApiRelation(serialize = SerializeType.ONLY_ID)
  UserResource user
}
```

So, I want to be be able to say
`/events/1`
to do a query against the events service, which gives us back events with a userId, so then the resulting json has relationships with 
`{ "data": { "id": "1", "type": "users" } }`

I also want to be able to say
`/events/1?include=user`
to do a query against the events service AND a query against the user service, so now i have a full user object to include, so a `"included": [ { "id": "1", "type": "users", "attributes": {...} } ]`

Even if there was a `LoopIncludeBehavior` of `Always call a relationship's findManyTargets or findOneTarget if the field is included`, that wouldn't be quite right, because I actually want to utilize the userId that I already have off of the event object, almost like a `/users/1` call, rather than do a query against the user service using the event id.

Any thoughts on how the framework wants me to go about doing this?

**Comment by mohlek** [comment]
Currently there is no way.
However with my latest changes it would be possible to implement on our own.
I hope to get this done soon and maybe get it upstream.

**Comment by remmeier** [comment]
totally agree with that. That @IdSettable setup would also allow to directly setup relationship repositories.

**Comment by remmeier** [comment]
maybe to clarify for non-insiders, it could then look like:

```
@JsonApiResource(type='events')
public class EventResource {
   @JsonApiRelation
   UserResource user
   
  @JsonApiRelationId
   String userId
}
```

The user attribute would then still be managed & looked-up by crnk automatically. There would be no need to speciy relationship repositories for such cases.

**Comment by mohlek** [comment]
I was thinking of something like this:

```Java
@JsonApiResource(type='events')
public class EventResource {
   @JsonApiRelation( idField = "userId" )
   UserResource user

   String userId
}
```

**Comment by remmeier** [comment]
the possiblity to specify the name manually sounds good to me. maybe "xyId" could  be the default.

**Comment by buschtoens** [comment]
I love @mohlek's code example. I have been desperately waiting for this feature since day one. It would unlock so many performance gains in our app :heart_eyes: :rocket:

**Comment by mohlek** [comment]
If the query set ```include user``` there should be a request to the relationship repository.

**Comment by mohlek** [comment]
> the possiblity to specify the name manually sounds good to me. maybe "xyId" could be the default.

So this means we don't need ```@JsonApiRelationId``` ?!
I don't see any need for this annotation if the field is referenced by `@JsonApiRelation`

**Comment by remmeier** [comment]
I think it is useful to have boths after experimenting with it. To explicitly mark with @JsonApiRelationId, to have a default naming convetion (like "Id" suffix) and do allow to override it with JsonApiRelation.idField if necessary. It makes things a bit more robust, meaning that some other field cannot change a field by referencing it as idField field, but it must be marked as one and so it is easy to decide if it is one or not.

**Comment by remmeier** [comment]
two commits found its way on the master branch. One introduced @JsonApiRelationId and @JsonApiRelation.idName. A second introduces @JsonApiRelation.repositoryBehavior that allows to setup a default relationship repository implementation, making a custom implementation unncessary (as long as the resource repositories on one/both end of the relationship obey some rules). Each of them can give quite dramatic improvement to the relationship handling. If you like you can start playing around. Please also give feedback. In a couple of days it could be released as Crnk 2.5. In the meantime you may want to build the Asciidoc documentation for more details (the online one is not updated so far).

**Comment by remmeier** [comment]
it is also not entirely unlikely that I missed something, as those things have an impact in many places within crnk-core. Altough existing behavior/annotations should be fairly safe from those changes.

**Comment by remmeier** [comment]
and special thx to @mohlek  for coming up with the relationship identifier thing

**Comment by remmeier** [comment]
yes, thx. That needs to be hardened. Will have it in a few minutes. It all needs to call findAll(ids, querySpec)

**Comment by remmeier** [comment]
is on master

**Comment by remmeier** [comment]
2.5.20180205212522

**Comment by jmchuster** [comment]
Thanks, that was way better than downloading the repo and mucking with the build.gradle.  Where do I see builds that aren't published on mvnrepository?

Issue I see here: https://github.com/crnk-project/crnk-framework/blob/master/crnk-core/src/main/java/io/crnk/core/repository/implicit/ImplicitOwnerBasedRelationshipRepository.java#L62

I get the exception `targetId=null not found for sourceId=1, field=user, sourceType=events`
When i access something like `/events/1/user` where the userId is null.

**Comment by remmeier** [comment]
I typically make use of checking it out and run publishToMavenLocal. Then you will have a 0.0.0-SNAPSHOT version avaialble in your local M2 repository. It is continuously published to oss.sonatype. But not sure whether it is possible to access it before promotion. Maybe some way.

yes, it should ignore does null. That needs to change. Anything else you find? I will also have https://github.com/crnk-project/crnk-framework/issues/216 available shortly.

**Comment by remmeier** [comment]
targetId null should now also be supported next to the new RelationshipMatcher class (see docs on master branch)

**Comment by jmchuster** [comment]
Nice.  I'm seeing some inconsistency in the casing of the queryParams (e.g. `filter[shared-event.userId EQ [1]]`).  I didn't spend enough time to nail down where this is coming from, and just worked around it by creating a camel-case/dash-case insensitive matcher.
Possibly related to https://github.com/crnk-project/crnk-framework/issues/196

**Comment by remmeier** [comment]
I added a comment to #196. And on master I did some further improvements to the relationship handling. There is now a ForwardingRelationshipRepository that is replacing RelationshipRepositoryBase and ImplicitXyRelationshipREpository. ForwardingRelationshipRepository takes two directions, OWNER and OPPOSITE to which repository to forward GET, UPDATE, DELETE and PATCH requests. GET and be set indepently from POST, DELETE, PATCH. This allows for four different combinations to cover different use cases. All in all this slowely concludes the entire relationship topic. Some more placing around to ensure everything works as it is supposed to.
